cranelift/codegen/src/ir/memflags.rs

```diff
@@ -326,32 +300,106 @@ impl MemFlags {
 
     /// Set the `checked` bit, returning new flags.
     pub const fn with_checked(self) -> Self {
-        self.with(FlagBit::Checked)
+        self.with_bit(BIT_CHECKED)
     }
 
-    /// Get the trap code to report if this memory access traps. If the
-    /// `notrap` flag is set, then return `None`. Otherwise, the choice
-    /// of trap code depends on the `tabletrap` flag. The default trap
-    /// code is `HeapOutOfBounds`, but with `tabletrap` set, the trap
-    /// code is instead `TableOutOfBounds`.
+    /// Get the trap code to report if this memory access traps.
+    ///
+    /// A `None` trap code indicates that this memory access does not trap.
     pub const fn trap_code(self) -> Option<TrapCode> {
-        if self.notrap() {
-            None
-        } else if self.tabletrap() {
-            Some(TrapCode::TableOutOfBounds)
-        } else {
-            Some(TrapCode::HeapOutOfBounds)
+        match (self.bits & MASK_TRAP_CODE) >> TRAP_CODE_OFFSET {
+            0b0000 => Some(TrapCode::HeapOutOfBounds), // default mode
+            0b0001 => Some(TrapCode::StackOverflow),
+            0b0010 => Some(TrapCode::HeapOutOfBounds),
```

Comment:
I'm inclined to think that we don't need to track whether `heap_oob` was set explicitly or just used as a default, so my instinct is to use just one encoding for it. I could easily be talked into the approach you've taken here, but I think it at least needs a bit more of a comment explaining why it appears twice here, and why the zero case doesn't appear in `with_trap_code` below.

Reply:
Nah I think that makes the most sense in retrospect too, no need to be needlessly complicated.